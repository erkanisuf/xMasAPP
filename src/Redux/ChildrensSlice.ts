import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { IProduct } from "../components/ChildComp/ChildComp";

import type { RootState } from "./store";

interface IReduxChildrenSlice {
  ChildrenApprovedItems: OBJtoAPI[];
  ChildrenDiscardedItems: OBJtoAPI[];
  myCart: IProduct[];
}
export interface OBJtoAPI {
  userId: number;
  date: Date | string;
  products: IProduct[];
}
// This is the Initial State represents the Products from the Open APi
const initialState: IReduxChildrenSlice = {
  ChildrenApprovedItems: [],
  ChildrenDiscardedItems: [],
  myCart: [],
};

export const ChildrensSlice = createSlice({
  name: "childrenSlice",
  //Redux Info
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    //Redux Tips
    // Use the PayloadAction type to declare the contents of `action.payload`
    ChildProductToApproved: (state, action: PayloadAction<OBJtoAPI>) => {
      const copyApprovedState = [...state.ChildrenApprovedItems];
      const findIndex = copyApprovedState.findIndex(
        (el) => el.userId === action.payload.userId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex !== -1) {
        copyApprovedState[findIndex].products.push(action.payload.products[0]);
      } else {
        copyApprovedState.push(action.payload);
      }

      state.ChildrenApprovedItems = copyApprovedState;
    },
    ChildProductToDiscarded: (state, action: PayloadAction<OBJtoAPI>) => {
      const copyDiscardedState = [...state.ChildrenDiscardedItems];
      const findIndex = copyDiscardedState.findIndex(
        (el) => el.userId === action.payload.userId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex !== -1) {
        copyDiscardedState[findIndex].products.push(action.payload.products[0]);
      } else {
        copyDiscardedState.push(action.payload);
      }

      state.ChildrenDiscardedItems = copyDiscardedState;
    },
    AddItemToMyCart: (state, action: PayloadAction<IProduct>) => {
      const copyMyCart = [...state.myCart];
      const product = { productId: action.payload.productId, quantity: 1 }; //New Object because the APi has random values of quantity
      const findIndex = copyMyCart.findIndex(
        (el) => el.productId === action.payload.productId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex !== -1) {
        copyMyCart[findIndex].quantity++;
      } else {
        copyMyCart.push(product);
      }

      state.myCart = copyMyCart;
    },
  },
});

export const {
  ChildProductToApproved,
  ChildProductToDiscarded,
  AddItemToMyCart,
} = ChildrensSlice.actions;

//Redux Setup from their Docs.
// Other code such as selectors can use the imported `RootState` type
export const selectCount = (state: RootState) => state;

export default ChildrensSlice.reducer;
