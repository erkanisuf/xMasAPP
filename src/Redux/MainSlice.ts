import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { IProductDesc } from "../components/Product/Product";
import type { RootState } from "./store";

interface IReduxMainSlice {
  products: IProductDesc[];
  myChildren: { id: number; name: string; image: string }[];
}
interface IChangePrice {
  id: number;
  discount: number;
}

// This is the Initial State represents the Products from the Open APi
const initialState: IReduxMainSlice = {
  products: [],
  myChildren: [
    {
      id: 1,
      name: "Pekka",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSVY3efFyktSBShjhBcNUt2JuH3F1r0qnlt1g&usqp=CAU",
    },
    {
      id: 2,
      name: "Mikko",
      image:
        "https://www.pngitem.com/pimgs/m/424-4240458_one-ok-rock-stand-out-fit-in-from.png",
    },
    {
      id: 3,
      name: "Juha",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRdBkq9ItXGCTqTIW4BdBlaEpZQVllKv8gkQGpdhVwfbAIuj9-4yOzapTjdNE0r14wSJoE&usqp=CAU",
    },
    {
      id: 4,
      name: "Jonna",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQlCT3cJxdz3-ntu8ixFRexArdaknweGrBQrw&usqp=CAU",
    },
    {
      id: 5,
      name: "Erkku",
      image:
        "https://st4.depositphotos.com/18690434/20926/v/450/depositphotos_209266078-stock-illustration-boy-kid-avatar-icon-vector.jpg",
    },
  ],
};

export const MainSlice = createSlice({
  name: "main",
  //Redux Info
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    //Redux Tips
    // Use the PayloadAction type to declare the contents of `action.payload`
    addProductsFromApi: (state, action: PayloadAction<IProductDesc[]>) => {
      state.products = action.payload;
    },
    changeProductPrice: (state, action: PayloadAction<IChangePrice>) => {
      const copyProducts = [...state.products];
      const findIndex = copyProducts.findIndex(
        (el) => el.id === action.payload.id
      );
      //Discount if more than 1 quantity
      if (findIndex !== -1) {
        copyProducts[findIndex] = {
          ...copyProducts[findIndex],
          discount: action.payload.discount,
        };
      }
      state.products = copyProducts;
    },
  },
});

export const { addProductsFromApi, changeProductPrice } = MainSlice.actions;

//Redux Setup from their Docs.
// Other code such as selectors can use the imported `RootState` type
export const selectCount = (state: RootState) => state;

export default MainSlice.reducer;
