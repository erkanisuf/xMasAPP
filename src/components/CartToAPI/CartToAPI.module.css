.containerCombined {
  display: flex;
  width: 100%;
}
.containerGreen {
  display: flex;
  font-size: 10px;
  width: 80% !important;
  flex-direction: column;

  flex: 1 !important;
  padding: 15px;
  background-color: #abf3db;
  text-align: center;
  font-size: 15px;
  margin: 5px;
  border-radius: 10px;
}
.containerRed {
  display: flex;
  font-size: 10px;
  width: 80% !important;
  flex-direction: column;

  flex: 1 !important;
  padding: 15px;
  background-color: #fdcfd1;
  text-align: center;
  font-size: 15px;
  margin: 5px;
  border-radius: 10px;
}
.containerRed h1 {
  font-size: 12px;
  width: 100%;
}
.containerGreen h1 {
  font-size: 12px;
}
.containerGreen img {
  width: 30%;
}
.containerRed img {
  width: 30%;
}

.closeButton {
  border: none;
  outline: none;
  cursor: pointer;
  font-size: 20px !important;
  color: #c57c7e;

  border: 3px solid #c57c7e;
  background-color: white;
  border-radius: 10px;
  padding: 10px;
  margin: 10px 0px;
  align-self: flex-end;
}

.closeButton:hover {
  color: #c45054;
}

.success {
  color: green;
  font-weight: 800;
  border: 1px solid green;
  padding: 5px;
  border-radius: 10px;
  background-color: white;
  margin: 5px auto;
  width: 100%;
}

.error {
  color: rgb(165, 14, 52);
  font-weight: 800;
  border: 1px solid rgb(165, 14, 52);
  padding: 5px;
  border-radius: 10px;
  background-color: white;
  margin: 5px auto;
  width: 100%;
}
