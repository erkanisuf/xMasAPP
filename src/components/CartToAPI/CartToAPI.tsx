import React, { useEffect, useState } from "react";

import { OBJtoAPI } from "../../Redux/ChildrensSlice";
import ChildrenImage from "../ChildrenImage/ChildrenImage";
import SmallSpinner from "../Spinner/SmallSpinner";
import CartToAPICSS from "./CartToAPI.module.css";

interface ICartToAPI {
  usercart: OBJtoAPI;
  type: TypeContainer;
}
interface Iitem {
  id: number;
  _id: string;
  products: any; // APi returns empty arr of this sinc eits not saving to the DB
}
type TypeContainer = "approved" | "discarded";
const CartToAPI: React.FC<ICartToAPI> = ({ usercart, type }) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [item, setItem] = useState<Iitem>({ id: 0, _id: "", products: [] });
  const [error, setError] = useState<boolean>(false);

  useEffect(() => {
    const abortCont = new AbortController();
    const postToAPI = () => {
      setLoading(true);
      fetch("https://fakestoreapi.com/carts", {
        method: "POST",
        signal: abortCont.signal,
        body: JSON.stringify(usercart),
      })
        .then((res) => res.json())
        .then((json) => {
          setItem(json);
          setLoading(false);
        })
        .catch((err) => {
          if (err.name !== "AbortError") {
            setError(true);
          }
        });
    };
    postToAPI();
    return () => abortCont.abort();
  }, [usercart]);

  if (loading) {
    return (
      <div
        className={
          type === "approved"
            ? CartToAPICSS.containerGreen
            : CartToAPICSS.containerRed
        }
      >
        <ChildrenImage name={usercart.userId} />
        <SmallSpinner />
      </div>
    );
  } else if (error) {
    return (
      <div
        className={
          type === "approved"
            ? CartToAPICSS.containerGreen
            : CartToAPICSS.containerRed
        }
      >
        <ChildrenImage name={usercart.userId} />
        <span className={CartToAPICSS.error}>Error , not saved!</span>
      </div>
    );
  }
  return (
    <div
      className={
        type === "approved"
          ? CartToAPICSS.containerGreen
          : CartToAPICSS.containerRed
      }
    >
      <ChildrenImage name={usercart.userId} />
      <span className={CartToAPICSS.success}>successfully saved!</span>

      <span>
        {" "}
        <span style={{ fontWeight: 800 }}>ID:</span>
        {item._id}
      </span>
    </div>
  );
};

export default CartToAPI;
