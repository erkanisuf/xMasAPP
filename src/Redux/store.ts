import { configureStore } from "@reduxjs/toolkit";
import ChildrensSlice from "./ChildrensSlice";
import MainSlice from "./MainSlice";
// ...

//MAIN: holds the products and the fake array of childrens object
// CHILDREN : holds the approved items and discarded items of the child , laso myCart property that holds the item that customer will  "buy"
export const store = configureStore({
  reducer: { main: MainSlice, childrens: ChildrensSlice },
});

// Redux required setup from Their Docs
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
