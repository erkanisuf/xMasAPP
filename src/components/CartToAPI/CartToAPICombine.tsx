import React from "react";
import { useAppSelector } from "../../Redux/hooks";
import CartToAPI from "./CartToAPI";
import CartToAPICSS from "./CartToAPI.module.css";
import { CgSmileMouthOpen } from "react-icons/cg";
import { CgSmileSad } from "react-icons/cg";
import { CgCloseR } from "react-icons/cg";
interface ICartToAPICombine {
  setOpen: any;
}
const CartToAPICombine: React.FC<ICartToAPICombine> = ({ setOpen }) => {
  const approvedItems = useAppSelector(
    (state) => state.childrens.ChildrenApprovedItems
  );
  const discardedItems = useAppSelector(
    (state) => state.childrens.ChildrenDiscardedItems
  );
  return (
    <div
      style={{
        width: "100%",
        height: "80%",
        display: "flex ",
        justifyContent: "flex-start",
        alignItems: "flex-start",
        flexDirection: "column",
        position: "relative",
      }}
    >
      <button
        className={CartToAPICSS.closeButton}
        onClick={() => setOpen(false)}
      >
        <CgCloseR />
        Close
      </button>
      <div
        style={{
          width: "100%",
          height: "80%",
          display: "flex ",
          justifyContent: "flex-start",
          alignItems: "flex-start",

          position: "relative",
        }}
      >
        <div
          style={{
            width: "100%",
            display: "flex",
            flexDirection: "column",
          }}
        >
          <h2
            style={{
              backgroundColor: "rgb(14, 165, 39)",
              width: "100%",
              color: "white",
              padding: "10px 0px",
              borderRadius: "10px",
              margin: "5px",
              textAlign: "center",
            }}
          >
            <CgSmileMouthOpen />
            Approved Items
          </h2>
          {approvedItems.map((el) => {
            return (
              <CartToAPI key={el.userId} usercart={el} type={"approved"} />
            );
          })}
        </div>
        <div
          style={{ width: "100%", display: "flex", flexDirection: "column" }}
        >
          <h2
            style={{
              backgroundColor: "rgb(165, 14, 52)",
              width: "100%",
              color: "white",
              margin: "5px",
              padding: "10px 0px",
              borderRadius: "10px",
              textAlign: "center",
            }}
          >
            <CgSmileSad />
            Discarded Items
          </h2>
          {discardedItems.map((el) => {
            return (
              <CartToAPI key={el.userId} usercart={el} type={"discarded"} />
            );
          })}
        </div>
      </div>
    </div>
  );
};

export default CartToAPICombine;
