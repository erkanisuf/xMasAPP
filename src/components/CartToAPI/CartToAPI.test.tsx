import React from "react";
import { Provider } from "react-redux";

import renderer from "react-test-renderer";
import { store } from "../../Redux/store";
import CartToAPI from "./CartToAPI";
import CartToAPICombine from "./CartToAPICombine";

it("snapshot <CartToAPI/> Component", () => {
  const testItem = {
    userId: 1,
    date: new Date(),
    products: [],
  };

  const tree = renderer
    .create(
      <Provider store={store}>
        <CartToAPI usercart={testItem} type={"approved"} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("snapshot <CartToAPICombine/> Component", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <CartToAPICombine setOpen={() => console.log("Test")} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
