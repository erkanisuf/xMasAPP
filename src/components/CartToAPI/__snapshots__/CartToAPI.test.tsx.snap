// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot <CartToAPI/> Component 1`] = `
<div
  className="containerGreen"
>
  <div
    className="containerChild"
  >
    <img
      alt=""
      src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSVY3efFyktSBShjhBcNUt2JuH3F1r0qnlt1g&usqp=CAU"
    />
    <h1>
      Pekka
      \`s wishlist 
    </h1>
  </div>
  <span
    className="success"
  >
    successfully saved!
  </span>
  <span>
     
    <span
      style={
        Object {
          "fontWeight": 800,
        }
      }
    >
      ID:
    </span>
    
  </span>
</div>
`;

exports[`snapshot <CartToAPICombine/> Component 1`] = `
<div
  style={
    Object {
      "alignItems": "flex-start",
      "display": "flex ",
      "flexDirection": "column",
      "height": "80%",
      "justifyContent": "flex-start",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <button
    className="closeButton"
    onClick={[Function]}
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16.3956 7.75734C16.7862 8.14786 16.7862 8.78103 16.3956 9.17155L13.4142 12.153L16.0896 14.8284C16.4802 15.2189 16.4802 15.8521 16.0896 16.2426C15.6991 16.6331 15.0659 16.6331 14.6754 16.2426L12 13.5672L9.32458 16.2426C8.93405 16.6331 8.30089 16.6331 7.91036 16.2426C7.51984 15.8521 7.51984 15.2189 7.91036 14.8284L10.5858 12.153L7.60436 9.17155C7.21383 8.78103 7.21383 8.14786 7.60436 7.75734C7.99488 7.36681 8.62805 7.36681 9.01857 7.75734L12 10.7388L14.9814 7.75734C15.372 7.36681 16.0051 7.36681 16.3956 7.75734Z"
        fill="currentColor"
      />
      <path
        clipRule="evenodd"
        d="M4 1C2.34315 1 1 2.34315 1 4V20C1 21.6569 2.34315 23 4 23H20C21.6569 23 23 21.6569 23 20V4C23 2.34315 21.6569 1 20 1H4ZM20 3H4C3.44772 3 3 3.44772 3 4V20C3 20.5523 3.44772 21 4 21H20C20.5523 21 21 20.5523 21 20V4C21 3.44772 20.5523 3 20 3Z"
        fill="currentColor"
        fillRule="evenodd"
      />
    </svg>
    Close
  </button>
  <div
    style={
      Object {
        "alignItems": "flex-start",
        "display": "flex ",
        "flexDirection": "column",
        "height": "80%",
        "justifyContent": "flex-start",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <h2
        style={
          Object {
            "backgroundColor": "rgb(14, 165, 39)",
            "borderRadius": "10px",
            "color": "white",
            "margin": "5px",
            "padding": "10px 0px",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 17C14.2091 17 16 15.2091 16 13H8C8 15.2091 9.79086 17 12 17Z"
            fill="currentColor"
          />
          <path
            d="M10 10C10 10.5523 9.55228 11 9 11C8.44772 11 8 10.5523 8 10C8 9.44772 8.44772 9 9 9C9.55228 9 10 9.44772 10 10Z"
            fill="currentColor"
          />
          <path
            d="M15 11C15.5523 11 16 10.5523 16 10C16 9.44772 15.5523 9 15 9C14.4477 9 14 9.44772 14 10C14 10.5523 14.4477 11 15 11Z"
            fill="currentColor"
          />
          <path
            clipRule="evenodd"
            d="M22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12ZM20 12C20 16.4183 16.4183 20 12 20C7.58172 20 4 16.4183 4 12C4 7.58172 7.58172 4 12 4C16.4183 4 20 7.58172 20 12Z"
            fill="currentColor"
            fillRule="evenodd"
          />
        </svg>
        Approved Items
      </h2>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <h2
        style={
          Object {
            "backgroundColor": "rgb(165, 14, 52)",
            "borderRadius": "10px",
            "color": "white",
            "margin": "5px",
            "padding": "10px 0px",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 11C9.55228 11 10 10.5523 10 10C10 9.44772 9.55228 9 9 9C8.44772 9 8 9.44772 8 10C8 10.5523 8.44772 11 9 11Z"
            fill="currentColor"
          />
          <path
            d="M14 17C14 15.8954 13.1046 15 12 15C10.8954 15 10 15.8954 10 17H8C8 14.7909 9.79086 13 12 13C14.2091 13 16 14.7909 16 17H14Z"
            fill="currentColor"
          />
          <path
            d="M16 10C16 10.5523 15.5523 11 15 11C14.4477 11 14 10.5523 14 10C14 9.44772 14.4477 9 15 9C15.5523 9 16 9.44772 16 10Z"
            fill="currentColor"
          />
          <path
            clipRule="evenodd"
            d="M22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12ZM20 12C20 16.4183 16.4183 20 12 20C7.58172 20 4 16.4183 4 12C4 7.58172 7.58172 4 12 4C16.4183 4 20 7.58172 20 12Z"
            fill="currentColor"
            fillRule="evenodd"
          />
        </svg>
        Discarded Items
      </h2>
    </div>
  </div>
</div>
`;
